### ISSUE TITLE
How to get auth code using msal

### ISSUE THREAD

**Issue by mikeyyyzhao** [AzureAD/microsoft-authentication-library-for-js#4931]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Wrapper Library

MSAL React (@azure/msal-react)

### Documentation Location

https://github.com/AzureAD/microsoft-authentication-library-for-js

### Description

is it possible to request auth code (https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow) using this library for React? If so, is it possible for someone to give me a code sample? I've tried looking around but have had a really difficult time getting the auth code. 

By using this [Azure sample](https://github.com/Azure-Samples/ms-identity-javascript-react-spa) and following [this tutorial](https://docs.microsoft.com/en-us/azure/active-directory/develop/tutorial-v2-react), I'm able to get the accessToken but can't figure out how to get the auth code (the response is returning null for `code: null`) to pass to our backend.

Currently, we're using this code snippet to show a popup and get the accessToken in the response. Is it possible to configure the code snippet below to only get the auth code so we can pass that to our backend to use and exchange for tokens?

```
instance.acquireTokenPopup(loginRequest).then(response => {
            }).catch(e => {
            })
```

**Comment by derisen** [comment]
@mikeyyyzhao msal-browser/msal-react does not expose the authorization code to the developer, and the authorization code obtained for your frontend shouldn't be passed to your backend. If you need to authenticate users to your backend, you should use [msal-node](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-node) instead. If you need to authenticate users on both frontend and backend, you might want to look into the hybrid spa auth code sample [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-browser-samples/HybridSample).

**Comment by logan-jobzmall** [comment]
@derisen how does one go about a situation such as this then: Angular SPA needs auth code to send to backend API server request so that the backend can then use that to request an Auth Token bundle, and then store those for later usage for offline access?

**Comment by derisen** [comment]
@logan-jobzmall can you provide more detail on your use case here? Tokens issued for a single page app have a short lifetime (1hr for access tokens, 24 hr for refresh tokens -[see](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/token-lifetimes.md)). It might be better to use a web app (confidential client) here with msal-node instead -the [hybrid spa code flow sample](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/token-lifetimes.md) shows how to authenticate a user at both backend and frontend.

A more complex approach could be to use the [OBO flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) in a web API to acquire tokens on-behalf-of the user that signed in to the SPA. In principle, you could refresh the tokens that the web API acquired continuously via a background worker and maintain offline access to another web API for longer periods of time. For more details, have a look at this [sample](https://github.com/Azure-Samples/ms-identity-dotnet-advanced-token-cache/blob/master/1-Integrated-Cache/1-2-WebAPI-BgWorker/README.md).

**Comment by derisen** [comment]
Closing this as resolved. Please let us know if you have a follow-up -thanks!

**Comment by xsegrity** [comment]
@derisen I have the same issue as @logan-jobzmall. I have an angular based spa app where I want the user to grant authorization for their personal Outlook Calendar so that my back end server can then add/update events on their behalf on an ongoing, non-interactive basis. This is easy in Google where I get an AccessCode on the client that I then send to the server to trade for an AccessToken/RefreshToken, then store the RefreshToken for later API usage without user interaction.

Is this even possible with MSAL/MS Graph?

**Comment by jrobertson-ed** [comment]
> @derisen I have the same issue as @logan-jobzmall. I have an angular based spa app where I want the user to grant authorization for their personal Outlook Calendar so that my back end server can then add/update events on their behalf on an ongoing, non-interactive basis. This is easy in Google where I get an AccessCode on the client that I then send to the server to trade for an AccessToken/RefreshToken, then store the RefreshToken for later API usage without user interaction.
> 
> Is this even possible with MSAL/MS Graph?

I have this exact issue, and not sure how to solve it..

**Comment by Orangeapplebot** [comment]
I have a web application with the same issue. My application has its own authentication functions. But, I want to add a feature of uploading files from onedrive. I want to get the authentication from my front end and store the tokens in backend along with the files to download.

**Comment by abdumanas-plutoflume** [comment]
> @derisen I have the same issue as @logan-jobzmall. I have an angular based spa app where I want the user to grant authorization for their personal Outlook Calendar so that my back end server can then add/update events on their behalf on an ongoing, non-interactive basis. This is easy in Google where I get an AccessCode on the client that I then send to the server to trade for an AccessToken/RefreshToken, then store the RefreshToken for later API usage without user interaction.


Have a similar requirement. Is this possible to do with `msal` libraries ?

**Comment by Alex-Cosma** [comment]
> @derisen I have the same issue as @logan-jobzmall. I have an angular based spa app where I want the user to grant authorization for their personal Outlook Calendar so that my back end server can then add/update events on their behalf on an ongoing, non-interactive basis. This is easy in Google where I get an AccessCode on the client that I then send to the server to trade for an AccessToken/RefreshToken, then store the RefreshToken for later API usage without user interaction.
> 
> Is this even possible with MSAL/MS Graph?

I am in the same situation and it is frustrating how difficult this is to do as compared to Google. Any success?

**Comment by abishekbaiju** [comment]
> > @derisen I have the same issue as @logan-jobzmall. I have an angular based spa app where I want the user to grant authorization for their personal Outlook Calendar so that my back end server can then add/update events on their behalf on an ongoing, non-interactive basis. This is easy in Google where I get an AccessCode on the client that I then send to the server to trade for an AccessToken/RefreshToken, then store the RefreshToken for later API usage without user interaction.
> > Is this even possible with MSAL/MS Graph?
> 
> I am in the same situation and it is frustrating how difficult this is to do as compared to Google. Any success?

have you found a solution? I still can't figure it out. the code is returned when I use the redirect method (it's present in the url). but when I use the popup login method the code is actually undefined.

**Comment by Alex-Cosma** [comment]
@abishekbaiju I actually managed to solve it, I don't even know if it's a ... hack? but it works. Check this out.

In your frontend app, you do something like this ('tis just test code)

<img width="960" alt="image" src="https://github.com/user-attachments/assets/19365b8c-cf24-4989-935d-80031c204fc3">

that "activeAccount" object contains this:

<img width="954" alt="image" src="https://github.com/user-attachments/assets/678a60d2-d918-437b-834f-fcd669725157">

that idToken is what I send to the backend, and I use the "onBehalfOf" functionality, like this:

<img width="667" alt="image" src="https://github.com/user-attachments/assets/9e091539-ebb6-46a6-bde9-162b4bbc13d5">


And it works. This is basically how I create the damn `GraphServiceClient` which I need.

I was lucky and it's stupidly documented, I'm not even sure if I did it right or not, but it just works.
